### PR TITLE
Allow different key sizes

### PR DESCRIPTION
**Pull request by idruzhitskiy**
Request for new structure of packages.
Closes #133

### CODE REVIEW

File: bob-common/src/data.rs [hunk 1]
```diff
@@ -8,7 +8,63 @@ use std::{
...
+#[derive(Copy, Clone, Hash, PartialEq, Eq, Debug)]
+pub struct BobKey([u8; BOB_KEY_SIZE]);
+
+impl From<Vec<u8>> for BobKey {
+    fn from(v: Vec<u8>) -> Self {
+        if v.len() > BOB_KEY_SIZE {
+            panic!("expected key size {}, received {}", BOB_KEY_SIZE, v.len());
+        }
+        let mut data = [0; BOB_KEY_SIZE];
+        for (ind, elem) in v.into_iter().enumerate() {
```
Comment (pyakushin) [inline bob-common/src/data.rs:22]: More efficient way to do this is via [`TryFrom`](https://doc.rust-lang.org/std/primitive.array.html#impl-TryFrom%3CVec%3CT%2C%20A%3E%3E) trait:
```
let data = v.try_into().expect("array/vec size mismatch");
```
To clone all buffer at once

**Review by pyakushin — APPROVED** [review]

File: bob-common/src/bob_client.rs [hunk 2]
```diff
@@ -71,7 +71,7 @@ pub mod b_client {
                 data: d.into_inner(),
             };
             let message = PutRequest {
-                key: Some(BlobKey { key }),
+                key: Some(BlobKey { key: key.into() }),
```
[the remark this answers is not in the document]
Comment (idruzhitskiy) [reply bob-common/src/bob_client.rs:74]: BlobKey is part of `.proto`, so there is no `From<Vec<u8>> BlobKey`

File: bob-common/src/mapper.rs [hunk 3]
```diff
@@ -136,12 +136,21 @@ impl Virtual {
...
-                .expect("u64 to u32"),
+                .expect("usize to u32"),
         }
     }
 
+    fn get_vdisk_id_by_mod(key: BobKey, len: usize) -> usize {
+        let mut rem = 0;
+        for byte in key.iter() {
+            rem += *byte as usize % len;
+            rem %= len;
```
Comment (ikopylov) [inline bob-common/src/mapper.rs:149]: Incorrect calculation of the remainder. Data written before the update is not readable because of this

Comment (ikopylov) [reply bob-common/src/mapper.rs:149]: C# proper implementation as an example:
```
            long rem = 0;
            long byteMaxRem = 1;
            for (int i = 0; i < _keyBytes.Length; i++)
            {
                rem = (rem + _keyBytes[i] * byteMaxRem) % divisor;
                byteMaxRem = (byteMaxRem * 256) % divisor;
            }

            return (int)rem;
```

Comment (ikopylov) [reply bob-common/src/mapper.rs:149]: This can be optimized by processing 4 bytes per iteration by converting to Int32

**Comment by ikopylov** [comment]
Do not merge. Something is wrong

**Review by ikopylov — CHANGES_REQUESTED** [review]
Byte ordering issue. Discussion required

**Comment by ikopylov** [comment]
On this version, sometimes the connection is canceled if the data is read by a non-existent key

File: bob-apps/bin/bobc.rs [hunk 4]
```diff
@@ -113,3 +114,13 @@ fn get_key_value(matches: &'_ ArgMatches<'_>) -> u64 {
...
         .expect("key must be u64")
 }
+
+fn get_key(k: u64) -> Vec<u8> {
+    // TODO move to static
+    let key_size = option_env!("BOB_KEY_SIZE")
+        .map_or(Ok(8), str::parse)
+        .expect("Could not parse BOB_KEY_SIZE");
+    let mut data = k.to_le_bytes().to_vec();
+    data.extend(repeat(0).take(key_size - data.len()));
```
Comment (ikopylov) [inline bob-apps/bin/bobc.rs:124]: It is faster and safer to do this inside `if` only when `key_size - data.len() > 0` (`key_size` in practice can be less than `data.len()`)

File: bob-apps/bin/bobp.rs [hunk 5]
```diff
@@ -116,6 +121,12 @@ impl TaskConfig {
     fn is_time_measurement_thread(&self) -> bool {
         self.measure_time
     }
+
+    fn get_proper_key(&self, key: u64) -> Vec<u8> {
+        let mut data = key.to_le_bytes().to_vec();
+        data.extend(repeat(0).take(self.key_size - data.len()));
```
Comment (ikopylov) [inline bob-apps/bin/bobp.rs:127]: Same here

File: bob-apps/bin/bobc.rs (hunk 4 above)
Comment (idruzhitskiy) [reply bob-apps/bin/bobc.rs]: Fixed

File: bob-apps/bin/bobp.rs (hunk 5 above)
Comment (idruzhitskiy) [reply bob-apps/bin/bobp.rs]: Fixed

File: bob-common/src/mapper.rs [hunk 6]
```diff
@@ -136,12 +136,18 @@ impl Virtual {
...
-            DistributionFunc::Mod => (key % self.vdisks.len() as u64)
+            DistributionFunc::Mod => (Self::get_vdisk_id_by_mod(key, self.vdisks.len()))
                 .try_into()
-                .expect("u64 to u32"),
+                .expect("usize to u32"),
         }
     }
 
+    fn get_vdisk_id_by_mod(key: BobKey, len: usize) -> usize {
+        key.iter().rev().fold([0, 1], |[rem, bmult], &byte| {
```
Comment (ikopylov) [inline bob-common/src/mapper.rs:146]: `rev()` should be removed for Little Endian

Comment (idruzhitskiy) [reply bob-common/src/mapper.rs]: Fixed

**Review by ikopylov — CHANGES_REQUESTED** [review]
Half of the data on a two-node cluster has stopped reading since the last change

Edit: this is expected (happened due to byte order change)

File: bob-apps/bin/bobc.rs [hunk 7]
```diff
@@ -113,3 +114,16 @@ fn get_key_value(matches: &'_ ArgMatches<'_>) -> u64 {
...
+fn get_key(k: u64) -> Vec<u8> {
+    // TODO move to static
+    let key_size = option_env!("BOB_KEY_SIZE")
+        .map_or(Ok(8), str::parse)
+        .expect("Could not parse BOB_KEY_SIZE");
+    let mut data = k.to_le_bytes().to_vec();
+    let diff = key_size - data.len();
+    if diff > 0 {
+        data.extend(repeat(0).take(diff));
+    }
```
Comment (ikopylov) [inline bob-apps/bin/bobc.rs:127]: If the diff is negative, then `data` length should be reduced

File: bob-apps/bin/bobp.rs [hunk 8]
```diff
@@ -116,6 +121,15 @@ impl TaskConfig {
     fn is_time_measurement_thread(&self) -> bool {
         self.measure_time
     }
+
+    fn get_proper_key(&self, key: u64) -> Vec<u8> {
+        let mut data = key.to_le_bytes().to_vec();
+        let diff = self.key_size - data.len();
+        if diff > 0 {
+            data.extend(repeat(0).take(diff));
+        }
```
Comment (ikopylov) [inline bob-apps/bin/bobp.rs:130]: Same here

File: bob-apps/bin/bobc.rs [hunk 9]
```diff
@@ -1,13 +1,22 @@
...
+
 use bob::{Blob, BlobKey, BlobMeta, BobApiClient, GetRequest, PutRequest};
 use clap::{App, Arg, ArgMatches, SubCommand};
 use http::Uri;
 use log::LevelFilter;
 use std::time::{SystemTime, UNIX_EPOCH};
 use tonic::Request;
 
+lazy_static! {
+    static ref KEY_SIZE: usize = option_env!("BOB_KEY_SIZE")
```
Comment (ikopylov) [inline bob-apps/bin/bobc.rs:15]: It is better to read `key_size` from command line arguments (like it is done in bobp)

File: bob-apps/bin/bobp.rs [hunk 10]
```diff
@@ -77,16 +76,34 @@ struct TaskConfig {
     payload_size: u64,
     direct: bool,
     measure_time: bool,
+    key_size: usize,
 }
 
 impl TaskConfig {
     fn from_matches(matches: &ArgMatches) -> Self {
+        let key_size = option_env!("BOB_KEY_SIZE")
```
Comment (ikopylov) [inline bob-apps/bin/bobp.rs:84]: It is better to read `key_size` from command line arguments

**Review by ikopylov — APPROVED** [review]
Looks good, except for the comments about reading `key_size` from command line arguments

File: bob-apps/bin/bobp.rs [hunk 11]
```diff
@@ -739,6 +737,14 @@ fn get_matches() -> ArgMatches<'static> {
                 .takes_value(false)
                 .long("verify"),
         )
+        .arg(
+            Arg::with_name("keysize")
+                .help("size of the binary key")
+                .takes_value(true)
+                .long("keysize")
+                .short("k")
+                .default_value("8"),
```
Comment (ikopylov) [inline bob-apps/bin/bobp.rs:746]: Can you fallback to `option_env!("BOB_KEY_SIZE")` if it is declared

**Pull request merged**